app/controllers/admin/dashboard_controller.rb

```diff
@@ -2,7 +2,7 @@ class Admin::DashboardController < AdminController
   before_action :logged_in_user, :admin_user
 
   def index
-    @posts = Post.order_updated_at
+    @posts = Post.order_updated_at.includes(:user)
```

Comment:
có những cách nào fix n+1 query e
sao e lại dùng includes?

Reply:
Có 3 cách dùng ạ :
Cách 1: Dùng eager_load:
- Thằng eager_load này sử dụng cơ chế của thằng select in() => Để khắc phục được lỗi N+1 query thì thằng select in() này chỉ cần 2 câu truy vấn thay vì N+1 câu truy vấn laf:
1. Truy vấn đầu tiên để load toàn bộ user
2. Truy vấn số 2 để load các post tương ứng với các users đó.
Cách 2: Dùng preload:
- Thằng preload này thì lại sử dụng cơ chế hoạt động của thằng joins => Để khắc phục tình trạng N+1 query thì thằng joins này sẽ thực hiện việc join các bảng trước rồi sau đó mới thực hiện truy vấn , vì vậy chỉ mất 1 câu lệnh truy vấn mà thôi thay vì N + 1 câu truy vấn.
Cách 3: Dùng includes:
- Thằng includes này thì có thể coi là thông minh nhất, mặc định thì nó sex dùng cơ chế thằng select in() nhưng khi cần thiết thì ta có thể thêm references vào để ép nó sang cơ chế thằng joins => nó có tính linh hoạt cao hơn !

Reply:
tìm hiểu lại nhé, đang bị sai đó

Reply:
anh ơi anh có thể nói cho e e sai ở đâu được không ạ :<

Reply:
em đã update lại comment rồi ạ

Reply:
nếu có where mà ko có references thì sao e

Reply:
Ví dụ : 
- User.all.includes(:posts).where("users.name = ?" , "Gate") => câu lệnh này được thực thi bình thường
- User.all.includes(:posts).where("posts.title  = ?" , "Hoa") => Câu lệnh này bị lỗi
=> Bởi vì select in() sẽ không load trước được dữ liệu từ bảng post, nên nó sẽ không hiểu trường title là gì . Còn trường name thì nằm trong table user rồi nên không gặp lỗi.
=> Để khắc phục lỗi này thì ta sử dụng cơ chế của joins ạ , khi hai bảng này được kết hợp với nhau thì luôn có đầy đủ thông tin các trường của mỗi bảng.
=> User.all.includes(:posts).references(:posts).where("posts.title  = ?" , "Hoa")